Extend odometry tests for steering controllers

AdityaPawar162 opened the pull request:
This PR addresses issue #1456 by:
1.Tracking odometry values before and after the timeout.
2.Ensuring position values remain stable.
3.Verifying that linear and angular velocities reset to zero.

christophfroehlich commented on steering_controllers_library/test/test_steering_controllers_library.cpp:
```diff
@@ -167,6 +189,14 @@ TEST_F(SteeringControllersLibraryTest, test_both_update_methods_for_ref_timeout)
   msg->twist.angular.z = TEST_ANGULAR_VELOCITY_Z;
   controller_->input_ref_.writeFromNonRT(msg);
 
+  // post odometry values
+  auto post_odom_x_2 = controller_->odometry_.get_x();
+  auto post_odom_y_2 = controller_->odometry_.get_y();
+
+  // Position should remain stable
+  EXPECT_NEAR(pre_odom_x_2, post_odom_x_2, position_tolerance);
+  EXPECT_NEAR(pre_odom_y_2, post_odom_y_2, position_tolerance);
+
```
this should go after the update() call?

christophfroehlich commented on steering_controllers_library/test/test_steering_controllers_library.cpp:
```diff
@@ -123,13 +123,27 @@ TEST_F(SteeringControllersLibraryTest, test_both_update_methods_for_ref_timeout)
+
+  const double position_tolerance = 1e-5;
+
+  // Position should remain stable
+  EXPECT_NEAR(pre_odom_x_1, post_odom_x_1, position_tolerance);
+  EXPECT_NEAR(pre_odom_y_1, post_odom_y_1, position_tolerance);
```
we should also test for the heading using `odometry_.get_heading()`

christophfroehlich reviewed (changes requested):
Thanks for giving a try! But apparently, this does not test the original issue. What I tried: Run colcon build and test with this version, and then with
`git revert 2375aca`
It succeeds in both cases.

AdityaPawar162 commented:
Hi @christophfroehlich,
I have updated the function as per suggested changes,I have also added a check for the `last_linear_velocity_` and `last_angular_velocity_`.

christophfroehlich commented:
tests are failing now

AdityaPawar162 commented:
> tests are failing now

Sorry , for the incorrect push.

AdityaPawar162 commented:
Hi @christophfroehlich ,
I have pushed the correct changes, please do suggest any changes to be made.

christophfroehlich reviewed (changes requested):
Thanks, the tests for the last*velocity now fails without the bugfix.

I had a look now why the other test does not fail: 
1.) The test is in open_loop=false mode
2.) The `update_odometry` call in `update_and_write_commands` is overriden in the `TestableSteeringControllersLibrary` and does not update the odometry class. 


We should add a new testfile with a different yaml (see add_rostest_with_parameters_gmock in the CMakeLists file) and add something like this in the TestableSteeringControllersLibrary (from bicycle_steering_controller)
```cpp
  if (params_.open_loop)
  {
    odometry_.update_open_loop(last_linear_velocity_, last_angular_velocity_, period.seconds());
  }
```

AdityaPawar162 commented:
> Thanks, the tests for the last*velocity now fails without the bugfix.
> 
> I had a look now why the other test does not fail: 1.) The test is in open_loop=false mode 2.) The `update_odometry` call in `update_and_write_commands` is overriden in the `TestableSteeringControllersLibrary` and does not update the odometry class.
> 
> We should add a new testfile with a different yaml (see add_rostest_with_parameters_gmock in the CMakeLists file) and add something like this in the `TestableSteeringControllersLibrary::update_odometry` (from bicycle_steering_controller)
> 
> ```c++
>   if (params_.open_loop)
>   {
>     odometry_.update_open_loop(last_linear_velocity_, last_angular_velocity_, period.seconds());
>   }
> ```

ok

AdityaPawar162 commented:
Hi @christophfroehlich,
As per your suggestion, I created a new YAML file with `open_loop` set to true and added a separate test file named `test_steering_controllers_open_library`. I also modified the `TestableSteeringControllersLibrary::update_geometry` function. However, I'm unsure why the builds are failing.

christophfroehlich commented on steering_controllers_library/CMakeLists.txt:
```diff
@@ -74,6 +74,20 @@ if(BUILD_TESTING)
+    controller_interface
+    hardware_interface
+  )
+
+  ament_add_gmock(test_steering_odometry_open_loop_timeout test/test_steering_odometry.cpp)
+  target_link_libraries(test_steering_odometry_open_loop_timeout steering_controllers_library)
+
```
```suggestion
```
why is this here?

AdityaPawar162 replied:
Oh, I thought I need to add another gmock test target.

christophfroehlich replied:
No, because this test file does not use the parameters at all?

christophfroehlich commented on steering_controllers_library/test/steering_controllers_library_open_loop_timeout_params.yaml:
```diff
@@ -0,0 +1,17 @@
+test_steering_controllers_library_open_loop_timeout:
```
```suggestion
test_steering_controllers_library:
```
this has to match the node name, which is hardcoded [here](https://github.com/ros-controls/ros2_controllers/blob/3850a08d73424963831a92bda2a9c332d8f0aa6f/steering_controllers_library/test/test_steering_controllers_library.hpp#L155)

christophfroehlich reviewed (changes requested):
See my other comments. With these changes, the test runs but fails with

```
3: /workspaces/ros2_rolling_ws/src/ros2_controllers/steering_controllers_library/test/test_steering_controllers_library.cpp:173: Failure
3: The difference between controller_->command_interfaces_[2].get_value() and 0.575875 is 0.071721078654024661, which exceeds 1e-6, where
3: controller_->command_interfaces_[2].get_value() evaluates to 0.50415392134597536,
3: 0.575875 evaluates to 0.57587500000000003, and
3: 1e-6 evaluates to 9.9999999999999995e-07.
3: 
3: /workspaces/ros2_rolling_ws/src/ros2_controllers/steering_controllers_library/test/test_steering_controllers_library.cpp:174: Failure
3: The difference between controller_->command_interfaces_[3].get_value() and 0.575875 is 0.092000372941643649, which exceeds 1e-6, where
3: controller_->command_interfaces_[3].get_value() evaluates to 0.66787537294164367,
3: 0.575875 evaluates to 0.57587500000000003, and
3: 1e-6 evaluates to 9.9999999999999995e-07.
3: 
3: /workspaces/ros2_rolling_ws/src/ros2_controllers/steering_controllers_library/test/test_steering_controllers_library.cpp:239: Failure
3: The difference between controller_->command_interfaces_[2].get_value() and 0.575875 is 0.071721078654024661, which exceeds 1e-6, where
3: controller_->command_interfaces_[2].get_value() evaluates to 0.50415392134597536,
3: 0.575875 evaluates to 0.57587500000000003, and
3: 1e-6 evaluates to 9.9999999999999995e-07.
3: 
3: /workspaces/ros2_rolling_ws/src/ros2_controllers/steering_controllers_library/test/test_steering_controllers_library.cpp:240: Failure
3: The difference between controller_->command_interfaces_[3].get_value() and 0.575875 is 0.092000372941643649, which exceeds 1e-6, where
3: controller_->command_interfaces_[3].get_value() evaluates to 0.66787537294164367,
3: 0.575875 evaluates to 0.57587500000000003, and
3: 1e-6 evaluates to 9.9999999999999995e-07.
```

I'm not sure why these are different in open_loop mode, but haven't investigated further now

christophfroehlich commented on steering_controllers_library/test/test_steering_controllers_library.cpp:
```diff
@@ -123,13 +123,33 @@ TEST_F(SteeringControllersLibraryTest, test_both_update_methods_for_ref_timeout)
   // age_of_last_command > ref_timeout_
   ASSERT_FALSE(age_of_last_command <= controller_->ref_timeout_);
   ASSERT_EQ((*(controller_->input_ref_.readFromRT()))->twist.linear.x, TEST_LINEAR_VELOCITY_X);
   ASSERT_EQ(
```
Another thing I realized: We have to call update() twice in the test, because the update_odometry() will be called in the beginning, but the last*velocity will be updated later. Calling update twice here, and we get also errors as (with the reverted fix)
```
3: /workspaces/ros2_rolling_ws/src/ros2_controllers/steering_controllers_library/test/test_steering_controllers_library.cpp:151: Failure
3: The difference between pre_odom_heading_1 and post_odom_heading_1 is 0.0030000000000000001, which exceeds position_tolerance, where
3: pre_odom_heading_1 evaluates to 0,
3: post_odom_heading_1 evaluates to 0.0030000000000000001, and
3: position_tolerance evaluates to 1.0000000000000001e-05.
```

AdityaPawar162 replied:
Called update function twice in the next commit.

christophfroehlich commented on steering_controllers_library/test/test_steering_controllers_library.cpp:
```diff
@@ -174,6 +203,19 @@ TEST_F(SteeringControllersLibraryTest, test_both_update_methods_for_ref_timeout)
     controller_->update(controller_->get_node()->now(), rclcpp::Duration::from_seconds(0.01)),
```
see above

AdityaPawar162 replied:
Called update function twice in the next commit.

AdityaPawar162 commented:
Hi @christophfroehlich,
The build is failing because it is not able to find ros2_control_cmake 
Starting >>> realtime_tools
Starting >>> control_msgs
Starting >>> ros2_control_test_assets
Starting >>> controller_manager_msgs
--- stderr: realtime_tools
CMake Error at CMakeLists.txt:4 (find_package):
  By not providing "Findros2_control_cmake.cmake" in CMAKE_MODULE_PATH this
  project has asked CMake to find a package configuration file provided by
  "ros2_control_cmake", but CMake did not find one.

  Could not find a package configuration file provided by
  "ros2_control_cmake" with any of the following names:

    ros2_control_cmakeConfig.cmake
    ros2_control_cmake-config.cmake

  Add the installation prefix of "ros2_control_cmake" to CMAKE_PREFIX_PATH or
  set "ros2_control_cmake_DIR" to a directory containing one of the above
  files.  If "ros2_control_cmake" provides a separate development package or
  SDK, be sure it has been installed.


---
Failed   <<< realtime_tools [0.74s, exited with code 1]
Aborted  <<< ros2_control_test_assets [3.00s]
Aborted  <<< controller_manager_msgs [6.3[9](https://github.com/ros-controls/ros2_controllers/actions/runs/14100888744/job/39496745904?pr=1606#step:7:10)s]
Aborted  <<< control_msgs [7.15s]

Summary: 0 packages finished [7.53s]
  1 package failed: realtime_tools
  3 packages aborted: control_msgs controller_manager_msgs ros2_control_test_assets
  1 package had stderr output: realtime_tools
  32 packages not processed
   
I believe there shouldn't be any installation errors in GitHub Actions, right?

AdityaPawar162 commented:
Hi @christophfroehlich,
I have made suggested changes, please do take a look.

christophfroehlich commented on steering_controllers_library/src/steering_controllers_library.cpp:
```diff
@@ -339,6 +339,15 @@ controller_interface::CallbackReturn SteeringControllersLibrary::on_activate(
+
+  for (size_t i = 0; i < steering_count_idx; i++)
+  {
+    command_interfaces_[steering_count_idx + i].set_value(0.575875);
```
?

AdityaPawar162 replied:
The test clearly expects the steering angles (command_interfaces_[2] and command_interfaces_[3]) to remain at 0.575875 during timeouts but somehow they were not so i directly set the steering angles to the expected value in the controller activation method (on_activate).

AdityaPawar162 replied:
I have added a timeout check in the `update_and_write_commands` function, ensuring that commands are updated only when the timeout is false.

christophfroehlich replied:
I don't think that this is the right approach. This value should be the steering angle for the requested twist, which should remain unchanged after timeout.

But something is strange, both steering angles should not be the same as this test uses ackermann config
https://github.com/ros-controls/ros2_controllers/blob/d1844e6da63f25d384a0d81bf8a48fdc9a7861d2/steering_controllers_library/test/test_steering_controllers_library.hpp#L125

christophfroehlich replied:
This answer refers to the line in the code of the thread.

christophfroehlich commented:
While checking the test class, I see that there are some unused/duplicate constants.

Could you clean them up as well please?
https://github.com/ros-controls/ros2_controllers/blob/d1844e6da63f25d384a0d81bf8a48fdc9a7861d2/steering_controllers_library/test/test_steering_controllers_library.hpp#L57-L61
https://github.com/ros-controls/ros2_controllers/blob/d1844e6da63f25d384a0d81bf8a48fdc9a7861d2/steering_controllers_library/test/test_steering_controllers_library.hpp#L292-L295
https://github.com/ros-controls/ros2_controllers/blob/d1844e6da63f25d384a0d81bf8a48fdc9a7861d2/steering_controllers_library/test/test_steering_controllers_library.hpp#L300-L309
https://github.com/ros-controls/ros2_controllers/blob/d1844e6da63f25d384a0d81bf8a48fdc9a7861d2/steering_controllers_library/test/test_steering_controllers_library.hpp#L311-L315
https://github.com/ros-controls/ros2_controllers/blob/d1844e6da63f25d384a0d81bf8a48fdc9a7861d2/steering_controllers_library/test/test_steering_controllers_library.hpp#L324

christophfroehlich commented on steering_controllers_library/src/steering_controllers_library.cpp:
```diff
@@ -386,32 +395,52 @@ controller_interface::return_type SteeringControllersLibrary::update_and_write_c
-
-    if (params_.front_steering)
+    // calculate new commands if not in timeout
+    if (!timeout)
```
I don't see the necessity of this change. 
In case of timeout it SHOULD set the appropriate commands see L397, L401 or L409/413, respectively.

AdityaPawar162 replied:
Actually, I specially wanted to update traction_commands and steering_commands when the timeout is false, so I thought there is a need to add this    
     `auto [traction_commands, steering_commands] = odometry_.get_commands(
      reference_interfaces_[0], reference_interfaces_[1], params_.open_loop,
      params_.reduce_wheel_speed_until_steering_reached);
    ` inside the flag.

AdityaPawar162 commented:
https://github.com/ros-controls/ros2_controllers/blob/d1844e6da63f25d384a0d81bf8a48fdc9a7861d2/steering_controllers_library/test/test_steering_controllers_library.hpp#L57-L61
This is getting used in the function `configure_odometry`
https://github.com/ros-controls/ros2_controllers/blob/d1844e6da63f25d384a0d81bf8a48fdc9a7861d2/steering_controllers_library/test/test_steering_controllers_library.hpp#L121-L128

christophfroehlich commented:
> This is getting used in the function `configure_odometry`

only FRONT_WHEELS_RADIUS_, WHEELBASE_, REAR_WHEEL_TRACK_

AdityaPawar162 commented:
Hi @christophfroehlich,
The `joint_command_values` are hardcoded to 2.2 and 4.4, I think so because of this the steering angle is not getting updated ?
https://github.com/ros-controls/ros2_controllers/blob/d1844e6da63f25d384a0d81bf8a48fdc9a7861d2/steering_controllers_library/test/test_steering_controllers_library.hpp#L318

AdityaPawar162 commented:
> Hi @christophfroehlich, The `joint_command_values` are hardcoded to 2.2 and 4.4, I think so because of this the steering angle is not getting updated ?
> 
> https://github.com/ros-controls/ros2_controllers/blob/d1844e6da63f25d384a0d81bf8a48fdc9a7861d2/steering_controllers_library/test/test_steering_controllers_library.hpp#L318

Hi @christophfroehlich,
Please clearify my doubt.

christophfroehlich commented:
@AdityaPawar162 I try to catch up and respond to your comments within the next days, could you please fix the conflicts with the master with a merge? (which come from #1166)

christophfroehlich commented:
@AdityaPawar162 are you still working on this?
